File: apps/api/src/f1hub/schemas/predictions.py
"""
Prediction Schemas

Pydantic models for ML prediction requests and responses.
"""

from typing import Any, Dict
from uuid import UUID

from pydantic import BaseModel, Field


class PredictionRequest(BaseModel):
    """Request to create a prediction."""

    model_name: str = Field(..., description="Name of the model to use")
    version: str = Field(default="latest", description="Model version")
    input_data: Dict[str, Any] = Field(..., description="Input features for prediction")


class TyreDegradationPredictionResponse(BaseModel):
    """Response from tyre degradation prediction."""

    stint_id: str
    predicted_deg_per_lap: float
    compound: str
    driver_id: str
    model_version: str


class LapTimePredictionResponse(BaseModel):
    """Response from lap time prediction."""

    predicted_lap_time: float
    tyre_age: int
    compound: str
    track_status: str
    position: int
    driver_id: str
    model_version: str


class OvertakePredictionResponse(BaseModel):
    """Response from overtake probability prediction."""

    overtake_probability: float
    gap_seconds: float
    closing_rate: float
    tyre_advantage: int
    drs_available: bool
    lap_number: int
    model_version: str


class RaceResultPredictionResponse(BaseModel):
    """Response from race result prediction."""

    predicted_position: int
    top3_probabilities: Dict[int, float]
    grid_position: int
    avg_lap_time: float
    driver_id: str
    model_version: str


class PredictionResponse(BaseModel):
    """Generic prediction response."""

    prediction_id: UUID | None = None
    model_name: str
    version: str
    prediction_value: Any
    confidence: Dict[str, float] | None = None

File: apps/web/src/types/predictions.ts
/**
 * Prediction Types
 *
 * TypeScript interfaces for ML predictions and models.
 */

export interface Model {
  id: string;
  model_name: string;
  version: string;
  model_type: string;
  status: string;
  metrics: Record<string, number>;
  artifact_path: string;
  created_at: string;
  updated_at: string;
}

export interface PredictionRequest {
  model_name: string;
  version: string;
  input_data: Record<string, any>;
}

export interface PredictionResult {
  prediction_id?: string;
  model_name: string;
  version: string;
  prediction_value: any;
  confidence?: Record<string, number>;
}

export interface TyreDegradationPrediction {
  stint_id: string;
  predicted_deg_per_lap: number;
  compound: string;
  driver_id: string;
  model_version: string;
}

export interface LapTimePrediction {
  predicted_lap_time: number;
  tyre_age: number;
  compound: string;
  track_status: string;
  position: number;
  driver_id: string;
  model_version: string;
}

export interface OvertakePrediction {
  overtake_probability: number;
  gap_seconds: number;
  closing_rate: number;
  tyre_advantage: number;
  drs_available: boolean;
  lap_number: number;
  model_version: string;
}

export interface RaceResultPrediction {
  predicted_position: number;
  top3_probabilities: Record<number, number>;
  grid_position: number;
  avg_lap_time: number;
  driver_id: string;
  model_version: string;
}

export type PredictionResponse =
  | TyreDegradationPrediction
  | LapTimePrediction
  | OvertakePrediction
  | RaceResultPrediction;
